File: dotnet/DesignPatterns/BehavioralDesignPatterns/VisitorDesignPattern/VisitorDesignPattern.Example1/ShoppingCart.cs
namespace VisitorDesignPattern.Example1;

/// <summary>
/// Shopping cart that acts as an Object Structure, holding a collection of products.
/// It allows visitors to traverse all products without exposing internal implementation.
/// </summary>
public class ShoppingCart
{
    private readonly List<IProduct> _products = new();

    public void AddProduct(IProduct product)
    {
        _products.Add(product);
        Console.WriteLine($"  Added: {product.Name} (${product.Price:F2})");
    }

    public decimal GetSubtotal()
    {
        return _products.Sum(p => p.Price);
    }

    public void Accept(IVisitor visitor)
    {
        foreach (var product in _products)
        {
            product.Accept(visitor);
        }
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/VisitorDesignPattern/VisitorDesignPattern.Example1/src/ShoppingCart.ts
import { IProduct } from './IProduct';
import { IVisitor } from './IVisitor';

export class ShoppingCart {
  private readonly _products: IProduct[] = [];

  addProduct(product: IProduct): void {
    this._products.push(product);
    console.log(`  Added: ${product.name} ($${product.price.toFixed(2)})`);
  }

  getSubtotal(): number {
    return this._products.reduce((sum, p) => sum + p.price, 0);
  }

  accept(visitor: IVisitor): void {
    for (const product of this._products) {
      product.accept(visitor);
    }
  }
}
